File: tests/unit/orbit_generation/magnetics/test_compute_magnetic_dipole.py
import sys
from pathlib import Path
import numpy as np
import pytest

ROOT = Path(__file__).resolve().parents[4]
sys.path.insert(0, str(ROOT / "build"))

import saltro_py


def test_dipole_dimensions_and_validity():
    N = 5
    Re = 6378136.3

    R = np.zeros((3, N))
    jtime = np.linspace(2451545.0, 2451545.0 + 0.01, N)

    for i in range(N):
        alt = 400e3
        r = Re + alt
        R[:, i] = [r, 0.0, 0.0]

    ok, B = saltro_py.compute_magnetic_dipole(R, jtime)

    assert ok
    assert B.shape == (3, N)

    assert np.all(np.isfinite(B))
    assert np.linalg.norm(B) > 0.0


def test_dipole_reasonable_magnitude_leo():
    N = 3
    Re = 6378136.3

    R = np.zeros((3, N))
    jtime = np.full(N, 2451545.0)

    alts = [200e3, 400e3, 700e3]

    for i in range(N):
        r = Re + alts[i]
        R[:, i] = [r, 0.0, 0.0]

    ok, B = saltro_py.compute_magnetic_dipole(R, jtime)
    assert ok

    Bmag = np.linalg.norm(B, axis=0)

    # Rough Earth dipole field magnitude ranges
    assert np.all(Bmag > 1e-7)
    assert np.all(Bmag < 1e-3)

    assert Bmag[0] > Bmag[1] > Bmag[2]


def test_dipole_circular_orbit_variation():
    N = 60
    Re = 6378136.3
    alt = 500e3
    r = Re + alt

    R = np.zeros((3, N))
    jtime = np.linspace(2451545.0, 2451545.0 + 0.01, N)

    for i in range(N):
        theta = 2.0 * np.pi * i / N
        R[:, i] = [
            r * np.cos(theta),
            r * np.sin(theta),
            0.0,
        ]

    ok, B = saltro_py.compute_magnetic_dipole(R, jtime)
    assert ok

    Bmag = np.linalg.norm(B, axis=0)

    assert np.all(np.isfinite(Bmag))
    assert np.all(Bmag > 0.0)

    # dipole variation shouldn't explode
    assert Bmag.max() / Bmag.min() < 50.0


def test_dipole_input_validation_shape():
    R = np.zeros((2, 5))
    jtime = np.ones(5)

    with pytest.raises(RuntimeError):
        saltro_py.compute_magnetic_dipole(R, jtime)


def test_dipole_input_validation_length():
    R = np.zeros((3, 5))
    jtime = np.ones(4)

    with pytest.raises(RuntimeError):
        saltro_py.compute_magnetic_dipole(R, jtime)

File: tests/unit/orbit_generation/magnetics/test_compute_magnetic_dipole.cpp
#include <catch2/catch_test_macros.hpp>
#include <Eigen/Dense>
#include <saltro/orbit_generation/magnetics/compute_magnetic_dipole.h>
#include <saltro/limits.h>
#include <cmath>

using namespace saltro;

TEST_CASE("MagneticDipole dimensions and validity", "[magnetic][dipole]")
{
    constexpr int N = 5;

    Eigen::Matrix<double,3,limits::MAX_LENGTH_TRAJ> R;
    Eigen::Matrix<double,1,limits::MAX_LENGTH_TRAJ> jtime;
    Eigen::Matrix<double,3,limits::MAX_LENGTH_TRAJ> B;

    const double Re = 6378136.3;

    for (int i = 0; i < N; ++i)
    {
        double alt = 400e3;
        double r = Re + alt;

        R.col(i) << r, 0.0, 0.0;
        jtime(i) = 2451545.0 + i * 1e-3;
    }

    bool ok = orbits::compute_magnetic_dipole(R, jtime, N, B);
    REQUIRE(ok);

    for (int i = 0; i < N; ++i)
    {
        REQUIRE(std::isfinite(B(0,i)));
        REQUIRE(std::isfinite(B(1,i)));
        REQUIRE(std::isfinite(B(2,i)));

        double mag = B.col(i).norm();
        REQUIRE(mag > 0.0);
    }
}


TEST_CASE("MagneticDipole order of magnitude sanity", "[magnetic][dipole]")
{
    constexpr int N = 3;

    Eigen::Matrix<double,3,limits::MAX_LENGTH_TRAJ> R;
    Eigen::Matrix<double,1,limits::MAX_LENGTH_TRAJ> jtime;
    Eigen::Matrix<double,3,limits::MAX_LENGTH_TRAJ> B;

    const double Re = 6378136.3;
    double alts[N] = {200e3, 400e3, 700e3};

    for (int i = 0; i < N; ++i)
    {
        double r = Re + alts[i];
        R.col(i) << r, 0.0, 0.0;
        jtime(i) = 2451545.0;
    }

    bool ok = orbits::compute_magnetic_dipole(R, jtime, N, B);
    REQUIRE(ok);

    double Bmag0 = B.col(0).norm();
    double Bmag1 = B.col(1).norm();
    double Bmag2 = B.col(2).norm();

    REQUIRE(Bmag0 > 1e-7);
    REQUIRE(Bmag0 < 1e-3);

    REQUIRE(Bmag1 > 1e-8);
    REQUIRE(Bmag1 < 1e-3);

    REQUIRE(Bmag2 > 1e-9);
    REQUIRE(Bmag2 < 1e-3);

    REQUIRE(Bmag0 > Bmag1);
    REQUIRE(Bmag1 > Bmag2);
}


TEST_CASE("MagneticDipole circular orbit smoothness", "[magnetic][dipole]")
{
    constexpr int N = 60;

    Eigen::Matrix<double,3,limits::MAX_LENGTH_TRAJ> R;
    Eigen::Matrix<double,1,limits::MAX_LENGTH_TRAJ> jtime;
    Eigen::Matrix<double,3,limits::MAX_LENGTH_TRAJ> B;

    const double Re = 6378136.3;
    const double alt = 500e3;
    const double r = Re + alt;

    for (int i = 0; i < N; ++i)
    {
        double theta = 2.0 * M_PI * i / N;

        R.col(i) << r * std::cos(theta),
                    r * std::sin(theta),
                    0.0;

        jtime(i) = 2451545.0 + i * 1e-3;
    }

    bool ok = orbits::compute_magnetic_dipole(R, jtime, N, B);
    REQUIRE(ok);

    double Bmin = 1e99;
    double Bmax = 0.0;

    for (int i = 0; i < N; ++i)
    {
        double mag = B.col(i).norm();
        REQUIRE(std::isfinite(mag));
        REQUIRE(mag > 0.0);

        Bmin = std::min(Bmin, mag);
        Bmax = std::max(Bmax, mag);
    }

    REQUIRE(Bmax / Bmin < 50.0); // dipole shouldn't explode
}